Sources/YUL/AST.swift

```diff
@@ -0,0 +1,312 @@
+//
+//  AST.swift
+//  YUL
+//
+//  Created by Yicheng Luo on 11/16/18.
+//
+
+import Foundation
+
+public typealias Identifier = String
+
+public enum Literal: CustomStringConvertible {
+  case num(Int)
+  case string_(String)
+  case bool_(Bool)
+  case decimal((n1: Int, n2: Int))
+  case hex(String)
+
+  public var description: String {
+    switch self {
+    case .num(let i):
+      return String(i)
+    case .string_(let s):
+      return "\"\(s)\""
+    case .bool_(let b):
+      return b ? "1" : "0"
+    case .decimal(let (n1, n2)):
+      return "\(n1).\(n2)"
+    case .hex(let h):
+      return h
+    }
+  }
+}
+
+public struct Block: CustomStringConvertible {
+  public var statements: [Statement]
+
+  public init(_ statements: [Statement]) {
+    self.statements = statements
+  }
+
+  public var description: String {
+    let statement_description = self.statements.map({ s in
+      return s.description
+    }).joined(separator: "\n")
+
+    return """
+    {
+      \(statement_description)
+    }
+    """
+  }
+
+}
+
+public struct If: CustomStringConvertible {
+  public let expression: Expression
+  public let block: Block
+
+  public init(_ expression: Expression, _ block: Block) {
+    self.expression = expression
+    self.block = block
+  }
+
+  public var description: String {
+    return "if \(expression.description) \(self.block)"
+  }
+}
+
+
+public struct Switch: CustomStringConvertible {
+  public let expression: Expression
+  public let cases: [(Literal, Block)]
+  public let default_: Block?
```

Comment:
What's with the random underscores here? If you want to use an identifier that's also a keyword then you should backtick escape it :)